cmd/erasure-zones.go

```diff
@@ -971,15 +1081,145 @@ func isTruncatedZones(zoneEntryChs [][]FileInfoCh, zoneEntries [][]FileInfo, zon
 	return isTruncated
 }
 
-func (z *xlZones) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string, maxKeys int) (ListObjectsInfo, error) {
-	if z.SingleZone() {
-		return z.zones[0].ListObjects(ctx, bucket, prefix, marker, delimiter, maxKeys)
+func isTruncatedZonesVersions(zoneEntryChs [][]FileInfoVersionsCh, zoneEntries [][]FileInfoVersions, zoneEntriesValid [][]bool) bool {
+	for i, entryChs := range zoneEntryChs {
+		for j := range entryChs {
+			zoneEntries[i][j], zoneEntriesValid[i][j] = entryChs[j].Pop()
+		}
 	}
 
+	var isTruncated = false
+	for _, entriesValid := range zoneEntriesValid {
+		for _, valid := range entriesValid {
+			if !valid {
+				continue
+			}
+			isTruncated = true
+			break
+		}
+		if isTruncated {
+			break
+		}
+	}
+	for i, entryChs := range zoneEntryChs {
+		for j := range entryChs {
+			if zoneEntriesValid[i][j] {
+				zoneEntryChs[i][j].Push(zoneEntries[i][j])
+			}
+		}
+	}
+	return isTruncated
+}
+
+func (z *erasureZones) listObjectVersions(ctx context.Context, bucket, prefix, marker, versionMarker, delimiter string, maxKeys int) (ListObjectVersionsInfo, error) {
+	loi := ListObjectVersionsInfo{}
+
+	if err := checkListObjsArgs(ctx, bucket, prefix, marker, z); err != nil {
+		return loi, err
+	}
+
+	// Marker is set validate pre-condition.
+	if marker != "" {
+		// Marker not common with prefix is not implemented. Send an empty response
+		if !HasPrefix(marker, prefix) {
+			return loi, nil
+		}
+	}
+
+	if marker == "" && versionMarker != "" {
+		return loi, NotImplemented{}
+	}
+
+	// With max keys of zero we have reached eof, return right here.
+	if maxKeys == 0 {
+		return loi, nil
+	}
+
+	// For delimiter and prefix as '/' we do not list anything at all
+	// since according to s3 spec we stop at the 'delimiter'
+	// along // with the prefix. On a flat namespace with 'prefix'
+	// as '/' we don't have any entries, since all the keys are
+	// of form 'keyName/...'
+	if delimiter == SlashSeparator && prefix == SlashSeparator {
+		return loi, nil
+	}
+
+	// Over flowing count - reset to maxObjectList.
+	if maxKeys < 0 || maxKeys > maxObjectList {
+		maxKeys = maxObjectList
+	}
+
+	if delimiter != SlashSeparator && delimiter != "" {
+		return loi, NotImplemented{}
+	}
+
+	// Default is recursive, if delimiter is set then list non recursive.
+	recursive := true
+	if delimiter == SlashSeparator {
+		recursive = false
+	}
+
+	var zonesEntryChs [][]FileInfoVersionsCh
+	var zonesEndWalkCh []chan struct{}
+
+	const ndisks = 3
```

Comment:
???

Above this is used as `if quorumCount < ndisks-1 {` - so only if `quorumCount < 2`?

Reply:
Yeah listing we decided only use 3 disks this is master not part of my change really.

Reply:
ah, ok, I can live with that.